### PR TITLE
chore(td-tools): enable eslint/strict-boolean-expressions and strictNullChecks

### PR DESCRIPTION
**Pull request by danielpeintner**
This PR addresses one checkmark in https://github.com/eclipse-thingweb/node-wot/issues/1046.

### CODE REVIEW

**Comment by relu91** [comment]
I started working on this, but I gave up. Thank you for bringing this up! I think we should pay double attention to the changes... because sometimes we used `null` as synonym of `undefined`

File: packages/td-tools/src/td-parser.ts [hunk 1]
```diff
@@ -222,20 +222,20 @@ export function serializeTD(thing: Thing): string {
     const copy = JSON.parse(JSON.stringify(thing));
 
     // clean-ups
-    if (!copy.security || copy.security.length === 0) {
+    if (copy.security === undefined || copy.security.length === 0) {
```
Comment (JKRhb) [inline packages/td-tools/src/td-parser.ts:225]:
```suggestion
    if (copy?.security.length === 0) {
```

File: packages/td-tools/src/td-parser.ts [hunk 2]
```diff
@@ -222,20 +222,20 @@ export function serializeTD(thing: Thing): string {
...
         copy.securityDefinitions = {
             nosec_sc: { scheme: "nosec" },
         };
         copy.security = ["nosec_sc"];
     }
 
-    if (copy.forms && copy.forms.length === 0) {
+    if (copy.forms !== undefined && copy.forms.length === 0) {
```
Comment (JKRhb) [inline packages/td-tools/src/td-parser.ts:232]:
```suggestion
    if (copy?.forms.length === 0) {
```

File: packages/td-tools/src/td-parser.ts [hunk 3]
```diff
@@ -265,11 +265,11 @@ export function serializeTD(thing: Thing): string {
             }
         }
     }
-    if (copy.events && Object.keys(copy.events).length === 0) {
+    if (copy.events !== undefined && Object.keys(copy.events).length === 0) {
         delete copy.events;
     }
 
-    if (copy.links && copy.links.length === 0) {
+    if (copy.links !== undefined && copy.links.length === 0) {
```
Comment (JKRhb) [inline packages/td-tools/src/td-parser.ts:272]:
```suggestion
    if (copy?.links.length === 0) {
```

File: packages/td-tools/src/thing-model-helpers.ts [hunk 4]
```diff
@@ -241,7 +241,7 @@ export class ThingModelHelpers {
             case "http": {
                 return new Promise((resolve, reject) => {
                     http.get(uri, (res) => {
-                        if (!res.statusCode || res.statusCode !== 200) {
+                        if (res.statusCode === undefined || res.statusCode !== 200) {
```
Comment (JKRhb) [inline packages/td-tools/src/thing-model-helpers.ts:244]:
```suggestion
                        if (res?.statusCode !== 200) {
```

File: packages/td-tools/src/thing-model-helpers.ts [hunk 5]
```diff
@@ -268,7 +268,7 @@ export class ThingModelHelpers {
                 return new Promise((resolve, reject) => {
                     https
                         .get(uri, (res) => {
-                            if (!res.statusCode || res.statusCode !== 200) {
+                            if (res.statusCode === undefined || res.statusCode !== 200) {
```
Comment (JKRhb) [inline packages/td-tools/src/thing-model-helpers.ts:271]:
```suggestion
                            if (res?.statusCode !== 200) {
```

File: packages/td-tools/src/thing-model-helpers.ts [hunk 6]
```diff
@@ -508,7 +508,7 @@ export class ThingModelHelpers {
                 extendedModel.properties = {};
             }
             for (const key in properties) {
-                if (dest.properties && dest.properties[key]) {
+                if (dest.properties && dest.properties[key] !== undefined) {
```
Comment (JKRhb) [inline packages/td-tools/src/thing-model-helpers.ts:511]:
```suggestion
                if (dest?.properties[key] !== undefined) {
```

File: packages/td-tools/src/thing-model-helpers.ts [hunk 7]
```diff
@@ -625,7 +625,7 @@ export class ThingModelHelpers {
         keys = keys.map((el) => el.replace("{{", "").replace("}}", ""));
         let isValid = true;
         let errors;
-        if (keys && keys.length > 0 && (map === undefined || map === null)) {
+        if (keys !== undefined && keys.length > 0 && (map === undefined || map === null)) {
```
Comment (JKRhb) [inline packages/td-tools/src/thing-model-helpers.ts:628]: With `foo == null`, you can actually check for `undefined` and `null` at the same time (from my knowledge, this is one of the few cases where `==` is valid to use).

```suggestion
        if ((keys ?? []).length > 0 && (map == null)) {
```

File: packages/td-tools/src/thing-model-helpers.ts [hunk 8]
```diff
@@ -376,7 +376,7 @@ export class ThingModelHelpers {
         if (!options) {
             options = {} as CompositionOptions;
         }
-        if (!options.baseUrl) {
+        if (options.baseUrl === undefined) {
             options.baseUrl = ".";
         }
```
Comment (JKRhb) [inline packages/td-tools/src/thing-model-helpers.ts:381]: Statements like these might actually be able to be simplified like this:

```suggestion
        options.baseUrl ??= ".";
```

Comment (danielpeintner) [reply packages/td-tools/src/thing-model-helpers.ts:381]: You are right, but to be honest I still have *issues* understanding the nullish coalescing operator at the first look. Is it just me? :-)
Shall we go over and update all of them?

**Comment by danielpeintner** [comment]
One more thing: Since there are lots of changes I am now more thinking about proposing PRs for each package step by step... otherwise it gets huge and hardly manageable.. 
What do people think?

**Comment by danielpeintner** [comment]
> I started working on this, but I gave up. Thank you for bringing this up! I think we should pay double attention to the changes... because sometimes we used `null` as synonym of `undefined`

I think after doing some more work on it I noticed this issue as well.

In a sense 

`if (bla !== undefined) {`

is not enough and we would almost need 

`if (bla !== undefined && bla !== null) {`

which I really don't like...

Ideas/Proposals?

**Comment by JKRhb** [comment]
> In a sense
> 
> `if (bla !== undefined) {`
> 
> is not enough and we would almost need
> 
> `if (bla !== undefined && bla !== null) {`
> 
> which I really don't like...
> 
> Ideas/Proposals?

I think in this case, you can (somewhat unintuitively) cover both with `bla != null`. When I integrated strict null checks in `node-coap`, that concept was very helpful in dealing with this scenario. As mentioned somewhere else, this is apparently one of the few scenarios where it is valid to use the `!=` operator instead of the `!==` operator.

File: packages/td-tools/src/thing-model-helpers.ts (hunk 4 above)
Comment (danielpeintner) [reply packages/td-tools/src/thing-model-helpers.ts]: I don't think the proposed change leads to the same result ---> reverting for now..

**Comment by danielpeintner** [comment]
@JKRhb I think some of the proposed changes are not correct and lead to issues. I suggest to keep them as is...
I think that's why the thing model helper tests are failing. I revert them ...

File: packages/td-tools/src/td-parser.ts [hunk 9]
```diff
@@ -222,20 +222,20 @@ export function serializeTD(thing: Thing): string {
     const copy = JSON.parse(JSON.stringify(thing));
 
     // clean-ups
-    if (!copy.security || copy.security.length === 0) {
+    if (copy?.security.length === 0) {
```
Comment (danielpeintner) [inline packages/td-tools/src/td-parser.ts:225]: I also think this proposed change is not correct ... since it is OR not AND

Comment (danielpeintner) [reply packages/td-tools/src/td-parser.ts]: fixed with https://github.com/eclipse-thingweb/node-wot/pull/1077/commits/25d5390b5c9213128cd51404367da2bc98f0f6d9

**Comment by danielpeintner** [comment]
@JKRhb @relu91 I am a bit lost and do not understand why some TM tests are failing. Can you spot the issue!?
I think I need a break 🙈

File: packages/td-tools/src/util/asset-interface-description.ts [hunk 10]
```diff
@@ -132,7 +132,7 @@ export class AssetInterfaceDescriptionUtil {
     public transformTD2SM(tdAsString: string, protocols?: string[]): string {
         const td: ThingDescription = TDParser.parseTD(tdAsString);
 
-        const aidID = td.id ? td.id : "ID" + Math.random();
+        const aidID = td.id != null ? td.id : "ID" + Math.random();
```
Comment (relu91) [inline packages/td-tools/src/util/asset-interface-description.ts:135]:
```suggestion
        const aidID = td.id ?? "ID" + Math.random();
```

File: packages/td-tools/src/util/asset-interface-description.ts [hunk 11]
```diff
@@ -295,12 +295,12 @@ export class AssetInterfaceDescriptionUtil {
         endpointMetadata?: Record<string, unknown>
     ): string | [string, ...string[]] {
         const security: string[] = [];
-        if (endpointMetadata?.value && endpointMetadata.value instanceof Array) {
+        if (endpointMetadata?.value != null && endpointMetadata.value instanceof Array) {
```
Comment (relu91) [inline packages/td-tools/src/util/asset-interface-description.ts:298]:
```suggestion
        if (endpointMetadata?.value instanceof Array) {
```

File: packages/td-tools/src/util/asset-interface-description.ts [hunk 12]
```diff
@@ -295,12 +295,12 @@ export class AssetInterfaceDescriptionUtil {
...
             for (const v of endpointMetadata.value) {
                 if (v.idShort === "security") {
-                    if (v.value && v.value instanceof Array) {
+                    if (v.value != null && v.value instanceof Array) {
```
Comment (relu91) [inline packages/td-tools/src/util/asset-interface-description.ts:301]:
```suggestion
                    if (v.value instanceof Array) {
```

File: packages/td-tools/src/util/asset-interface-description.ts [hunk 13]
```diff
@@ -400,12 +399,16 @@ export class AssetInterfaceDescriptionUtil {
...
+            submodel instanceof Object &&
+            submodel.idShort != null &&
+            submodel.idShort === "AssetInterfacesDescription"
+        ) {
+            if (submodel.submodelElements != null && submodel.submodelElements instanceof Array) {
                 for (const submodelElement of submodel.submodelElements) {
                     if (submodelElement instanceof Object) {
                         logDebug("SubmodelElement.idShort: " + submodelElement.idShort);
-                        if (submodelRegex && typeof submodelRegex === "string" && submodelRegex.length > 0) {
+                        if (submodelRegex != null && typeof submodelRegex === "string" && submodelRegex.length > 0) {
```
Comment (relu91) [inline packages/td-tools/src/util/asset-interface-description.ts:411]:
```suggestion
                        if (typeof submodelRegex === "string" && submodelRegex.length > 0) {
```

File: packages/td-tools/src/util/asset-interface-description.ts [hunk 14]
```diff
@@ -422,7 +425,7 @@ export class AssetInterfaceDescriptionUtil {
 
     private processSubmodelElement(smInformation: SubmodelInformation, submodelElement: Record<string, unknown>): void {
         // EndpointMetadata vs. InterfaceMetadata
-        if (submodelElement.value && submodelElement.value instanceof Array) {
+        if (submodelElement.value != null && submodelElement.value instanceof Array) {
```
Comment (relu91) [inline packages/td-tools/src/util/asset-interface-description.ts:428]:
```suggestion
        if (submodelElement.value instanceof Array) {
```

File: packages/td-tools/src/util/asset-interface-description.ts [hunk 15]
```diff
@@ -446,7 +449,7 @@ export class AssetInterfaceDescriptionUtil {
                 if (smValue instanceof Object) {
                     if (smValue.idShort === "InterfaceMetadata") {
                         logInfo("InterfaceMetadata");
-                        if (smValue.value && smValue.value instanceof Array) {
+                        if (smValue.value != null && smValue.value instanceof Array) {
```
Comment (relu91) [inline packages/td-tools/src/util/asset-interface-description.ts:452]:
```suggestion
                        if (smValue.value instanceof Array) {
```

File: packages/td-tools/src/thing-model-helpers.ts [hunk 16]
```diff
@@ -625,7 +625,7 @@ export class ThingModelHelpers {
         keys = keys.map((el) => el.replace("{{", "").replace("}}", ""));
         let isValid = true;
         let errors;
-        if (keys && keys.length > 0 && (map === undefined || map === null)) {
+        if (keys != null && keys.length > 0 && (map === undefined || map === null)) {
```
Comment (relu91) [inline packages/td-tools/src/thing-model-helpers.ts:628]:
```suggestion
        if (keys?.length > 0 && (map === undefined || map === null)) {
```

**Review by relu91 — CHANGES_REQUESTED** [review]
Some further simplifications.

File: packages/td-tools/src/thing-model-helpers.ts (hunk 8 above)
Comment (relu91) [reply packages/td-tools/src/thing-model-helpers.ts]: Is this conversation resolved? Are we going to enable the rule for eslint? if that's so we should have a tracking issue.

Comment (JKRhb) [reply packages/td-tools/src/thing-model-helpers.ts]: I just opened #1092, so this conversation can be resolved :)

**Comment by danielpeintner** [comment]
Note: I am not sure "why" but the CI gets stuck all the time in "Test with coverage report" with this PR. Local tests run fine...

**Comment by relu91** [comment]
> Note: I am not sure "why" but the CI gets stuck all the time in "Test with coverage report" with this PR. Local tests run fine...

I was surprised too, let's see if #1093 get stuck too.

File: packages/td-tools/src/td-parser.ts [hunk 17]
```diff
@@ -222,20 +222,20 @@ export function serializeTD(thing: Thing): string {
     const copy = JSON.parse(JSON.stringify(thing));
 
     // clean-ups
-    if (!copy.security || copy.security.length === 0) {
+    if (copy.security == null || copy.security.length === 0) {
```
Comment (JKRhb) [inline packages/td-tools/src/td-parser.ts:225]: I think one error in the CI is actually caused by this line:

```suggestion
    if (copy?.security.length === 0) {
```

Comment (danielpeintner) [reply packages/td-tools/src/td-parser.ts:225]: @JKRhb Can you explain why this should cause issues? Locally the tests are all "green".
I even think the change you are proposing has a different meaning.

The current code 
`    if (!copy.security || copy.security.length === 0) {`
enters the if branch IF either `copy.security` is `null` or `undefined ` but also IF the security array is of `length === 0`

Your code misses the first part, doesn't it ?

Comment (JKRhb) [reply packages/td-tools/src/td-parser.ts:225]: Hmm, yeah, you are right, I was mistaken with my suggestion, I thought the logic got reversed, but it should still be right. What got me to make the suggestion is that there is currently this error reported in the CI:

```
  1) the server side of servient
       should be able to add a Thing based on WoT.ThingFragment:
     TypeError: Cannot read properties of undefined (reading 'length')
      at Object.serializeTD (/home/runner/work/node-wot/node-wot/packages/td-tools/src/td-parser.ts:232:21)
      at ExposedThing.getThingDescription (src/exposed-thing.ts:126:30)
      at WoTServerTest.<anonymous> (test/ServerTest.ts:109:22)
      at Generator.next (<anonymous>)
      at fulfilled (test/ServerTest.ts:11:58)

```

However, I somehow made the suggestion to the wrong line. I would resolve this thread and try again 😅

File: packages/td-tools/src/td-parser.ts [hunk 18]
```diff
@@ -222,20 +222,20 @@ export function serializeTD(thing: Thing): string {
...
         copy.securityDefinitions = {
             nosec_sc: { scheme: "nosec" },
         };
         copy.security = ["nosec_sc"];
     }
 
-    if (copy.forms && copy.forms.length === 0) {
+    if (copy?.forms.length === 0) {
```
Comment (JKRhb) [inline packages/td-tools/src/td-parser.ts:232]: After also testing locally, I now found out that the real issue is in this line :)

```suggestion
    if (copy.forms?.length === 0) {
```

Comment (danielpeintner) [reply packages/td-tools/src/td-parser.ts:232]: Yes, that makes sense! Sorry for not seeing it myself.. I still wonder why it works locally on my side!?

File: packages/td-tools/src/thing-model-helpers.ts [hunk 19]
```diff
@@ -138,7 +138,7 @@ export class ThingModelHelpers {
     public static getModelVersion(data: ThingModel): string | undefined {
         if (
             "version" in data &&
-            data.version &&
+            data.version != null &&
             typeof data.version === "object" &&
             "model" in data.version &&
             typeof data.version.model === "string"
```
Comment (JKRhb) [inline packages/td-tools/src/thing-model-helpers.ts:144]: Maybe as a last-minute suggestion: I think these checks could be replaced with `typeof data?.version?.model === "string"` (or alternatively the whole method with `return data?.version?.model`).

Comment (JKRhb) [reply packages/td-tools/src/thing-model-helpers.ts]: I see, thank you for making the adjustment! :)

**Review by JKRhb — APPROVED** [review]

**Review by relu91 — APPROVED** [review]
Greenlight from my side too!

**Pull request merged**